### PR TITLE
Created AvroFileReader

### PR DESCRIPTION
**Pull request by Marianie-Simeon**
### **Addresses** 
https://github.com/broadinstitute/gatk/pull/7115

### **Commit Summary**
-Created AvroFileReader
- Update ExtractCohort and ExtractCohortEngine to accept a AvroFile

Testing:
Created a test for AvroFileReader

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortEngine.java [hunk 1]
```diff
@@ -90,7 +91,8 @@ public ExtractCohortEngine(final String projectID,
                                final double vqsLodINDELThreshold,
                                final ProgressMeter progressMeter,
                                final ExtractCohort.QueryMode queryMode,
-                               final String filterSetName) {
+                               final String filterSetName,
+                               final String Avrofilepath) {
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortEngine.java:95]: put this next to the table parameter (`cohortTableName`) and use a similar name like `cohortAvroFileName`

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortEngine.java [hunk 2]
```diff
@@ -111,6 +113,7 @@ public ExtractCohortEngine(final String projectID,
         this.queryMode = queryMode;
 
         this.filterSetName = filterSetName;
+        this.Avrofilepath = Avrofilepath;
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortEngine.java:116]: use java case style convention (ie avroFilePath or avroFilepath)

File: src/main/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReader.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.avro.Schema;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Iterator;
+
+public class AvroFileReader implements GATKAvroReader {
+
+        private DatumReader<GenericRecord> datumReader;
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReader.java:21]: indenting is off

File: src/main/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReader.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            throw new GATKException("Error closing stream", e);
+        }
+    }
+
+    @Override
+    public Iterator<GenericRecord> iterator() {
+        return this;
+    }
+
+    public static void main(String[] args){
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReader.java:73]: what's the main method for?

File: src/test/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReaderUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.util.*;
+
+/**
+ * A class to test the functionality of {@link AvroFileReader}.
+ */
+public class AvroFileReaderUnitTest extends GATKBaseTest {
+
+    private static final String avroFileURI = "gs://broad-dsp-spec-ops/kcibul/acmg.35.chr20.1mb.cohort.extract.avro";
+    private static final AvroFileReader AvroFile = new AvroFileReader(avroFileURI);
```
Comment (kcibul) [inline src/test/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReaderUnitTest.java:18]: java naming conventions would be `avroFile` for the variable name.  There's a bunch of these in here

File: src/test/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReaderUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    public void testAvroFileNext() {
+        GenericRecord AvroFileNext = AvroFile.next();
+        String  TestAvroFileNext  =   "{\"location\": 20000000060998, \"sample_name\": \"SM-GXZUY\", \"state\": \"0\", \"ref\": null, \"alt\": null, \"call_GT\": null, \"call_GQ\": null, \"call_RGQ\": null, \"QUALapprox\": null, \"AS_QUALapprox\": null, \"call_PL\": null}";
+        Assert.assertEquals(AvroFileNext.toString(), TestAvroFileNext, "AvroFile Next did not match.");
+    }
+
+    @Test()
+    public void testiterator() {
+        Iterator<GenericRecord> AvroFileTest = AvroFile.iterator();
+        Assert.assertEquals(AvroFile, AvroFileTest, "AvroFile did not match.");
```
Comment (kcibul) [inline src/test/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReaderUnitTest.java:43]: what is this testing?  you're comparing an iterator to an AvroFile?  what does that do?

**Review by kcibul — COMMENTED** [review]
Making good progress -- there are some java naming conventions that need to be cleaned up but the structure looks good.  Can you expand the avro option back out through ExtractCohort so we can run the tool with either a table OR a avro file?

File: src/main/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReader.java (hunk 4 above)
Comment (Marianie-Simeon) [reply src/main/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReader.java]: I left it in there by mistake, I used it to tested

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortEngine.java [hunk 7]
```diff
@@ -163,8 +167,14 @@ public void traverse() {
                     logger.debug("using storage api with local sort");
                 }
                 logger.debug("Initializing Reader");
-                final StorageAPIAvroReader storageAPIAvroReader = new StorageAPIAvroReader(cohortTableRef, rowRestriction, projectID);
-                createVariantsFromUngroupedTableResult(storageAPIAvroReader, fullVqsLodMap, fullYngMap, noFilteringRequested);
+                if (cohortAvroFileName == null){
+                    final StorageAPIAvroReader StorageAPIAvroReader = new StorageAPIAvroReader(cohortTableRef, rowRestriction, projectID);
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortEngine.java:171]: lower case the java variablename `final StorageAPIAvroReader storageAPIAvroReader`

File: src/main/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReader.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            throw new GATKException("Error closing stream", e);
+        }
+    }
+
+    @Override
+    public Iterator<GenericRecord> iterator() {
+        return this;
+    }
+
+    }
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReader.java:71]: This should have no indent, it closes the class

File: src/test/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReaderUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.avro.Schema;
+
+import java.util.*;
+
+/**
+ * A class to test the functionality of {@link AvroFileReader}.
+ */
+public class AvroFileReaderUnitTest extends GATKBaseTest {
+
+    private static final String avroFileURI = "gs://broad-dsp-spec-ops/kcibul/acmg.35.chr20.1mb.cohort.extract.avro";
```
Comment (kcibul) [inline src/test/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReaderUnitTest.java:17]: a few things here

1. make this file much smaller (~ kilobytes), just take the first 100 rows or something
2. move it into GATK repo (@mmorgantaylor has done this for other tests)

**Review by kcibul — APPROVED** [review]
Looks good -- just a few nits and one comment about test data that should be addressed

File: src/test/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReaderUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public class AvroFileReaderUnitTest extends GATKBaseTest {
+
+    private static final String avroFileName = "src/test/java/org/broadinstitute/hellbender/utils/bigquery/avro_test_avro_test_file.avro";
+    private static final AvroFileReader avroFile = new AvroFileReader(avroFileName);
+
+    @Test()
+    public void testGetSchema() {
+        Schema avroFileSchema = avroFile.getSchema();
+        System.out.println("avroFileSchema");
```
Comment (mmorgantaylor) [inline src/test/java/org/broadinstitute/hellbender/utils/bigquery/AvroFileReaderUnitTest.java:23]: are these print lines needed for the test? probably remove them (here and in testAvroFileNext below)

**Pull request closed without merging**